go-backend/db/user.go

```diff
@@ -2,21 +2,135 @@ package db
+	"time"
 
 	logger "github.com/sirupsen/logrus"
 )
 
+// User - basic struct representing a User
 type User struct {
-	Name string `db:"name" json:"full_name"`
-	Age  int    `db:"age" json:"age"`
+	ID              int           `db:"id" json:"id"`
+	Name            string        `db:"name" json:"full_name"`
+	OrgID           int           `db:"org_id" json:"org_id"`
+	Email           string        `db:"email" json:"email"`
+	DisplayName     string        `db:"display_name" json:"display_name"`
+	ProfileImageURL string        `db:"profile_image_url" json:"profile_image_url"`
+	SoftDelete      bool          `db:"soft_delete" json:"soft_delete"`
+	RoleID          int           `db:"role_id" json:"role_id"`
+	Hi5QuotaBalance int           `db:"hi5_quota_balance" json:"hi5_quota_balance"`
+	SoftDeleteBy    sql.NullInt64 `db:"soft_delete_by" json:"soft_delete_by"`
+	SoftDeleteOn    sql.NullTime  `db:"soft_delete_on" json:"soft_delete_on"`
+	CreatedAt       time.Time     `db:"created_at" json:"created_at"`
 }
 
+// Organization - retrieve the user's organization based on the OrgID property
+func (u *User) Organization(ctx context.Context, store Storer) (org Organization, err error) {
+	org, err = store.GetOrganization(ctx, u.OrgID)
+	return
+}
+
+// Role - retrieve the user's role based on the RoleID property
+func (u *User) Role(ctx context.Context, store Storer) (role Role, err error) {
+	role, err = store.GetRoleByID(ctx, u.RoleID)
+	return
+}
+
+// GetUserByEmail - Given an email address, return that user.
+func (s *pgStore) GetUserByEmail(ctx context.Context, email string) (user User, err error) {
+	err = s.db.Get(&user, `SELECT * FROM users WHERE email=$1 LIMIT 1`, email)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			err = ae.ErrRecordNotFound
+		}
+		// Possible that there's no rows in the result set
+		logger.WithField("err", err.Error()).Error("Error selecting user from database by email " + email)
+		return
+	}
+
+	// Populate the user's organization data
+	org := Organization{}
+	err = s.db.Get(&org, `SELECT * FROM organizations WHERE id=$1 LIMIT 1`, user.OrgID)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error selecting user's organization from database by email and id " + email + " " + string(user.OrgID))
+	}
+	return
+}
+
+// GetUserByID - Given the database ID for that user, look them up.
+func (s *pgStore) GetUserByID(ctx context.Context, id int) (user User, err error) {
+	err = s.db.Get(&user, `SELECT * FROM users WHERE id=$1 LIMIT 1`, id)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error selecting user from database by id " + string(id))
+		return
+	}
+	// Populate the user's organization data
+	org := Organization{}
+	err = s.db.Get(&org, `SELECT * FROM organizations WHERE id=$1`, user.OrgID)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error selecting user's organization from database by user id and org id " + string(id) + " " + string(user.OrgID))
```

Comment:
should return in error block

Reply:
010a355cbaf1bee655f9dbe74fb3fae70dc5fade